### PR TITLE
[Feature] Implement fast version of RTMDet.

### PR DESCRIPTION
**Pull request by RangiLyu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Implement a fast version of RTMDet. 
The training speed is 2.6 times faster than the previous version.

| version | training time of RTMDet-s | mAP|
| :------: | :--: | :--: |
| mmdet |  37h  |44.6 |
| mmyolo(fast) |  14h  | 44.6|

## Modification

1. Replace datasample with YOLOv5 data structure.
2. Refactor the loss and dynamic soft label assigner to support batch parallel computing.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by PeterH0323 — COMMENTED** [review]
建议 config 中将 `num_classes = 80` 改为变量设置，暴露给用户方便其进行修改

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 1]
```diff
@@ -183,7 +179,8 @@ def forward(self, feats: Tuple[Tensor, ...]) -> tuple:
             for reg_layer in self.reg_convs[idx]:
                 reg_feat = reg_layer(reg_feat)
 
-            reg_dist = self.rtm_reg[idx](reg_feat) * stride
+            # reg_dist = self.rtm_reg[idx](reg_feat) * stride
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/rtmdet_head.py:182]: delete

File: mmyolo/models/dense_heads/rtmdet_head.py [hunk 2]
```diff
@@ -418,286 +305,130 @@ def loss_by_feat(
...
+        pad_bbox_flag = (gt_bboxes.sum(-1, keepdim=True) > 0).float()
+
         device = cls_scores[0].device
-        anchor_list, valid_flag_list = self.get_anchors(
-            featmap_sizes, batch_img_metas, device=device)
+
+        # If the shape does not equal, generate new one
+        if featmap_sizes != self.featmap_sizes:
+            self.featmap_sizes = featmap_sizes
+            self.multi_level_anchors = self.prior_generator.grid_priors(
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/rtmdet_head.py:318]:
```python
self.featmap_sizes = featmap_sizes
mlvl_priors = self.prior_generator.grid_priors(
                featmap_sizes,
                dtype=cls_scores[0].dtype,
                device=cls_scores[0].device,
                with_stride=True)
self.flatten_priors = torch.cat(mlvl_priors, dim=0)
self.mlvl_priors = [mlvl[:, :2] for mlvl in mlvl_priors]
```
It needs to be replaced with the above writing method, otherwise there will be potential bugs. See https://github.com/open-mmlab/mmyolo/issues/373 for details

File: mmyolo/models/task_modules/assigners/batch_dsl_assigner.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+EPS = 1.0e-7
+
+
+@TASK_UTILS.register_module()
+class BatchDynamicSoftLabelAssigner(nn.Module):
+    """Computes matching between predictions and ground truth with dynamic soft
+    label assignment.
+
+    Args:
+        soft_center_radius (float): Radius of the soft center prior.
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_dsl_assigner.py:23]: `num_classes` docstr

**Review by hhaAndroid — APPROVED** [review]

**Review by hhaAndroid — APPROVED** [review]
